### DISCUSSION TITLE
Databricks PySpark Performance Issue

### DISCUSSION THREAD

**Discussion by ghost** [moj-analytical-services/splink#1567] [Q&A]
Hello, 

I've noticed that inclusion of latitude-longitude columns significantly slows down my computation in Databricks and PySpark (to the point where operations haven't even completed) regardless of if I include a blocker or comparison which includes these columns. 

For context, I'm running record linkage only for two Spark dataframes of 2mil and 6mil rows respectively. I select the following columns, all of which are strings except for latitude and longitude, which are floats rounded up to the thousandth place.

`cols = ['first_name', 'middle_initial', 'last_name', 'city', 'area_code_0', 'phone_number_0', 'specialty_code', 'unique_id', 'zip_code', 'latitude', 'longitude']`

When I run `linker.estimate_u_using_random_sampling(max_pairs=1e7)`, my tasks run, but after about 1 minute, the tasks pause on one job and show no progress. The program still runs without failing, but the progress bar doesn't move. (If I remove latitude and longitude from my columns, this program runs in 3 minutes.)

I've looked into the Optimising Spark Performance page and have tweaked these settings, but to no avail.

I'm wondering if anyone else has encountered this issue before? Is it specifically an issue with FloatType data?

Thank you in advance,
Jackson

**Comment by RobinL** [comment]
I've played about with this an I don't see an obvious performance bottleneck.

Here's some code I wrote to see whether the formula used for distance is inefficient.

(Note I obtained the code that Splink used to calculate distance from 
```
import splink.spark.comparison_level_library as cll
print(cll.distance_in_km_level("lat", "lng", 10).as_dict()["sql_condition"])`
```
)

<details><summary>benchmarking code</summary>

```python
import random
import time

from pyspark.context import SparkConf, SparkContext
from pyspark.sql import SparkSession

conf = SparkConf()


conf.set("spark.driver.memory", "10g")
conf.set("spark.sql.shuffle.partitions", "8")

sc = SparkContext.getOrCreate(conf=conf)

spark = SparkSession(sc)


data = [(random.uniform(-90, 90), random.uniform(-180, 180)) for _ in range(3000)]

schema = ["lat", "lng"]
df = spark.createDataFrame(data, schema=schema)


df.createOrReplaceTempView("df")

sql = """
select l.lat as lat_l, l.lng as lng_l, r.lat as lat_r, r.lng as lng_r
from df as l
cross join df as r
"""

df_joined = spark.sql(sql)

df_joined.createOrReplaceTempView("df_joined")

df_joined.cache()


for i in range(3):
    # sql1 is the original formula, in sql2 and sql3 i look to see whether alternatives are significantly faster.  They don't seem to be .
    sql1 = """
    select cast(
                acos(

            case
                when (
            sin( radians(`lat_l`) ) * sin( radians(`lat_r`) ) +
            cos( radians(`lat_l`) ) * cos( radians(`lat_r`) )
                * cos( radians(`lng_r` - `lng_l`) )
        ) > 1 then 1
                when (
            sin( radians(`lat_l`) ) * sin( radians(`lat_r`) ) +
            cos( radians(`lat_l`) ) * cos( radians(`lat_r`) )
                * cos( radians(`lng_r` - `lng_l`) )
        ) < -1 then -1
                else (
            sin( radians(`lat_l`) ) * sin( radians(`lat_r`) ) +
            cos( radians(`lat_l`) ) * cos( radians(`lat_r`) )
                * cos( radians(`lng_r` - `lng_l`) )
        )
            end

                ) * 6371
                as float
            )
        <= 10 as match

    from df_joined
    """

    # Timing
    n_runs = 10
    total_time = 0
    for _ in range(n_runs):
        start_time = time.time()
        spark.sql(sql1).toPandas()
        elapsed_time = time.time() - start_time
        total_time += elapsed_time

    average_time = total_time / n_runs
    print(f"Average runtime: {average_time} seconds")


    sql2 = """
    select cast(
                acos(

            (
            sin( radians(`lat_l`) ) * sin( radians(`lat_r`) ) +
            cos( radians(`lat_l`) ) * cos( radians(`lat_r`) )
                * cos( radians(`lng_r` - `lng_l`) )
            )   


                ) * 6371
                as float
            )
        <= 10 as match

    from df_joined
    """

    # Timing
    n_runs = 10
    total_time = 0
    for _ in range(n_runs):
        start_time = time.time()
        spark.sql(sql2).toPandas()
        elapsed_time = time.time() - start_time
        total_time += elapsed_time

    average_time = total_time / n_runs
    print(f"Average runtime: {average_time} seconds")


    sql3 = """
    select 
        sqrt(
            pow(`lat_l` - `lat_r`, 2) + pow(`lng_l` - `lng_r`, 2)
        ) <= 10 as match
    from df_joined
    """


    # Timing
    n_runs = 10
    total_time = 0
    for _ in range(n_runs):
        start_time = time.time()
        spark.sql(sql3).toPandas()
        elapsed_time = time.time() - start_time
        total_time += elapsed_time

    average_time = total_time / n_runs
    print(f"Average runtime: {average_time} seconds")

```
</details>


Could you share your other settings?  Are all of the other columns exact match columns, or are you using e.g. string similarity measures?  

You could try repartitioning the data prior to loading it into Splink, but I don't think that's particularly likely to be the source of the issue.

```
linker.estimate_u_using_random_sampling(max_pairs=1e7)
```

What happens if you set max pairs to `1e6 `?  Does it work then?  (Again, I don't think this should be much of an issue, we commonly use 1e8 or greater)

What happens if you run the benchmarking code above (you might need to change the spark setup)?  That should create about 10 million comparisons, so computationally it should be similar to what Splink is doing

**Comment by ghost** [comment]
Apologies for such a late response, especially after your extremely prompt one. I was working on a tight deadline and shifting away from lat-long km distance to a city then state CLL did the trick very well.

My location comparison was as follows:

`postcode_comparison = {
    'output_column_name': 'zip_code',
    'comparison_description': 'Postcode',
    'comparison_levels': [
        cll.null_level("zip_code"),
        cll.exact_match_level("zip_code"),
        cll.distance_in_km_level("latitude", "longitude", 1),
        cll.distance_in_km_level("latitude", "longitude", 10),
        cll.distance_in_km_level("latitude", "longitude", 50),
        cll.else_level()
    ],
}`

and my overall settings are listed here: 

`settings = {
    "link_type": "link_only",
    "comparisons": [
        ctl.name_comparison('first_name'),
        ctl.name_comparison('last_name', term_frequency_adjustments=True),
        postcode_comparison,
        cl.exact_match("middle_initial"),
        cl.exact_match("phone_number_0")
    ],
    "blocking_rules_to_generate_predictions": [
        block_on(["first_name", "last_name"]),
        block_on(["substr(first_name, 1, 1)", "last_name", "zip_code"]),
    ],
    "retain_matching_columns": False,
    "retain_intermediate_calculation_columns": False,
    "probability_two_random_records_match": 1.53e-07
}`

I tested with latitude and longitude as floats and as strings to confirm it wasn't a type issue.

Sharing if it's helpful for you and something you'd like to test more, but as I said, on my end, the other columns worked quite well. Even reducing max_pairs to `1e6` never led to completion, but now I can run `1e10` in less than 10 minutes without the lat-long comparison.

Thank you again for your response, and thank you for the excellent library!

**Reply by RobinL** [reply]
thank you!
